### PR TITLE
Tweak reads to not require a recordId

### PR DESCRIPTION
**Pull request by LiranCohen**
From https://github.com/TBD54566975/dwn-sdk-js/issues/468

This is a first pass, currently it accepts either `recordId` or `protocol` with `protocolPath` for a `RecordsRead`.
If the latest record is disallowed by protocol rules or not implicitly/explicitly allowed it will fail with a 401.

There is a potential to add additional fields to the description such as author, recipient, published etc for further filtering here.

Will think about how this fits in with https://github.com/TBD54566975/dwn-sdk-js/issues/467

### CODE REVIEW

**Comment by diehuxx** [comment]
I'm not sure about this feature. This feels like we're conflating query and read. Especially given your comment @LiranCohen about possible adding "additional fields to the description such as author, recipient, published, etc." I added a comment in https://github.com/TBD54566975/dwn-sdk-js/issues/468.

**Comment by LiranCohen** [comment]
After yesterday's office hours discussion I think this is ready to be reviewed.

As I was learning how protocols work, I had context + parent mixed up so i changed allowing to issue a Read based on a parentId along with a path instead of just the path.

Path based RecordsRead will look for the most recent Record which has not been Deleted. If no records are found at all it will 404.

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [inline src/interfaces/records-read.ts:86]: What is the use case for supporting `parentId`?

File: json-schemas/interface-methods/records-read.json
Comment (thehenrytsai) [inline json-schemas/interface-methods/records-read.json:52]: Sanity: the `allOf` outer wrapping doesn't appear to be necessary?

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [inline src/interfaces/records-read.ts:78]: The ugly types like `Partial<Omit<RecordsReadDescriptor, 'method'>>` would typically suggest improvement or refactoring somewhere.

Perhaps this can be improved after the merge of #474

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [inline src/interfaces/records-read.ts:76]:
```suggestion
   * @throws {DwnError} if `recordId` and either `protocol` or `protocolPath` are missing
```

File: tests/interfaces/records-read.spec.ts
Comment (thehenrytsai) [inline tests/interfaces/records-read.spec.ts:50]: Should we make it such that `recordId` and `protocol + protocolPath` are mutually exclusive?

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [inline src/interfaces/records-read.ts:79]: Nit:
```suggestion
    const commonFilter: Filter = { interface: DwnInterfaceName.Records, isLatestBaseState: true };
```

File: src/interfaces/records-read.ts
Comment (LiranCohen) [reply src/interfaces/records-read.ts:86]: Was thinking this would be useful if you have a protocol with multiple children paths that you may want to read a record from.

Say you have a game protocol with multiple `game` paths and you'd like to read the latest `game/score` from a particular game, you would need that game's `parentId`. You could read the score of the latest updated game without it, or target a specific game.

As I mentioned in the Issue discussion https://github.com/TBD54566975/dwn-sdk-js/issues/468#issuecomment-1682174577, I'd be open to removing this if it doesn't fit the intended purpose of this change.

File: src/interfaces/records-read.ts
Comment (LiranCohen) [reply src/interfaces/records-read.ts:78]: Yeah, I really didn't love that either.

And yup! Looking at #474, it would eliminate this from being used outside the context of `RecordsRead`.

File: json-schemas/interface-methods/records-read.json
Comment (LiranCohen) [reply json-schemas/interface-methods/records-read.json]: Thanks for catching this, I removed it.

I was trying to have the schema validation respond with a message that was more accurate rather than `recordId` is mossing, but it became too verbose and was not much better in messaging.

Simplified here, but replied separately to the testing in another comment.

File: src/interfaces/records-read.ts
Comment (LiranCohen) [reply src/interfaces/records-read.ts]: Yeah this was worded poorly by me, ended up going with:
```
@throws {DwnError} when either `recordId` is missing, or both `protocol` and `protocolPath` are missing.
```

I think that's a better description.

File: tests/interfaces/records-read.spec.ts
Comment (LiranCohen) [reply tests/interfaces/records-read.spec.ts]: Yup, that's actually how the behavior is expected, but these tests were structured/worded poorly.

I've gone ahead and restructured the tests to better represent the intention, but this brought out that the error returned by the JSON Schema validator always asks for a `recordId`, which isn't really descriptive enough.

I was thinking about adding a specific DwnError here for create and failing fast before it even gets down to the Schema validation. This way the error can be more descriptive similar to the one thrown when receiving a bad RecordsRead during handle.

I'll still keep the rules that exist within the JSON Schema, but would just throw a better error before it gets there.

Comment (LiranCohen) [reply tests/interfaces/records-read.spec.ts]: I ended up doing this and added a DwnError: `RecordsReadMissingCreateProperties`

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [reply src/interfaces/records-read.ts]: Let's remove it unless @csuwildcat have other thoughts. My understanding is that the whole point of this feature is to avoid needing to learning a CID that requires prior query, a scenario that needs `parentId` seems to defeat that entire purpose thus is out of scope.

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [reply src/interfaces/records-read.ts:88]: Sounds good, let's resolve the conflicts now that #474 is in, I'll do a final pass on the PR!

File: json-schemas/interface-methods/records-read.json
Comment (thehenrytsai) [inline json-schemas/interface-methods/records-read.json:48]: Should we do `oneOf`?

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [inline src/interfaces/records-read.ts:97]: Sanity: don't think we need this check, as JSON schema check would have caught this condition prior to this function call, so this is a condition only triggered by test.

Consider removing this condition, or if you prefer this for defensive programming, rename the error to something like `DwnErrorCode.RecordsReadCreateFilterUnreachableCode`

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [inline src/interfaces/records-read.ts:85]: Let's move it into `handlers/reads-read.ts` now that this is only used by the handler.

File: tests/handlers/records-read.spec.ts
Comment (thehenrytsai) [inline tests/handlers/records-read.spec.ts:822]:
```suggestion
          it('returns the latest RecordsWrite to the specified protocolPath when multiple children exist in path', async () => {
```

File: src/handlers/records-read.ts
Comment (thehenrytsai) [inline src/handlers/records-read.ts:42]: With this PR, the returned messages could be in 100s or even 1000s, and is subject to sort/paging. What were the scenario that prompted this change again? If it's for path that will ever have only one record, should we add a check for it?

File: tests/interfaces/records-read.spec.ts
Comment (diehuxx) [inline tests/interfaces/records-read.spec.ts:48]: What is the `P` at the end of `protocolPathOnlyP` and `protocolOnlyP`?

File: src/handlers/records-read.ts
Comment (diehuxx) [inline src/handlers/records-read.ts:43]: This will return the most recently _updated_ record, not the most recently created record. Is this the expected behavior?

File: tests/handlers/records-read.spec.ts
Comment (diehuxx) [inline tests/handlers/records-read.spec.ts:1385]:
```suggestion
      it('should reject with 400 if neither `recordId` not `protocol`+`protocolPath` are set', async () => {
```

File: src/handlers/records-read.ts
Comment (diehuxx) [inline src/handlers/records-read.ts:92]: Why `public`? If it's just for testing, I'd prefer to make this private, since testing this separately feels like testing implementation rather than behavior.

File: tests/handlers/records-read.spec.ts
Comment (diehuxx) [inline tests/handlers/records-read.spec.ts:1938]: extra whitespace here

File: tests/handlers/records-read.spec.ts
Comment (diehuxx) [inline tests/handlers/records-read.spec.ts:1952]: extra whitespace here too

File: tests/handlers/records-read.spec.ts
Comment (diehuxx) [inline tests/handlers/records-read.spec.ts:1960]: Any reason why you're using `['']` notation instead of dot notation here? I prefer dot notation since the typechecker isn't good with square bracket notation.

File: tests/handlers/records-read.spec.ts
Comment (diehuxx) [inline tests/handlers/records-read.spec.ts:1921]: Sanity: This method never throws. Why test that it doesn't throw under these conditions?

File: tests/handlers/records-read.spec.ts
Comment (diehuxx) [inline tests/handlers/records-read.spec.ts:1933]: Same as https://github.com/TBD54566975/dwn-sdk-js/pull/470/files#r1299587478
Sanity: This method never throws. Why test that it doesn't throw under these conditions?

File: tests/interfaces/records-read.spec.ts
Comment (diehuxx) [inline tests/interfaces/records-read.spec.ts:42]: Nit: Rather than just testing that it doesn't reject, can you check that `recordId` is present once it resolves?

File: tests/interfaces/records-read.spec.ts
Comment (diehuxx) [inline tests/interfaces/records-read.spec.ts:1]: Please add a test for when all of `recordId`, `protocol`, and `protocolPath` are present.

File: tests/handlers/records-read.spec.ts
Comment (LiranCohen) [reply tests/handlers/records-read.spec.ts:1960]: Filter's type is generalized to `[property: string]: ...` so there is no strong typing of any of the filter properties.

File: src/handlers/records-read.ts
Comment (LiranCohen) [reply src/handlers/records-read.ts:42]: The scenario is just to get the latest record under the path, but it doesn't have to be a "$singleton" record, we are moving away from singleton and toward $keep instead, which could have a value greater than 1.

So I don't think there are any checks here to provide, however we would add a limit to 1 and sort appropriately when that is exposed.

File: tests/interfaces/records-read.spec.ts
Comment (LiranCohen) [reply tests/interfaces/records-read.spec.ts:48]: The `P` stands for `Promise`, I usually put that as a naming convention when I haven't awaited the promise.

I can avoid that if you think it adds confusion (or use the full word).

File: src/handlers/records-read.ts
Comment (LiranCohen) [reply src/handlers/records-read.ts:43]: Good call out, this was my understanding of the intended behavior, do you think that's unclear?

When you think about using this feature, do you find yourself thinking of scenarios that latest Created should be returned instead of updated?

File: src/handlers/records-read.ts
Comment (LiranCohen) [reply src/handlers/records-read.ts:92]: It originally lived on RecordsRead interface and was used in multiple places before the Grants change, since it now lives within the handler and is only used here this could be private.

File: tests/handlers/records-read.spec.ts
Comment (LiranCohen) [reply tests/handlers/records-read.spec.ts:1933]: Henry made an update removing the explicitly thrown message here since JSON Schema will throw an exception elsewhere.

I'll update these tests to reflect that better.

File: tests/interfaces/records-read.spec.ts
Comment (thehenrytsai) [reply tests/interfaces/records-read.spec.ts:60]: I'd go for full word.

File: src/interfaces/records-read.ts
Comment (thehenrytsai) [reply src/interfaces/records-read.ts]: Hey @LiranCohen, yeah I understand the intent, but as it was, it was just for tests to hit the code, the actual running of DWN would never hit it thus I removed it.

As mentioned above, if you want to have the check in the same place, better to rename the error. If you want to move it so that it is actually called at runtime, then totally open to hear your proposal, but I suspect it will be a much more involved discussion.

File: src/handlers/records-read.ts
Comment (thehenrytsai) [reply src/handlers/records-read.ts:42]: > The scenario is just to get the latest record under the path

This sounds more like a "requirement" but not scenario to me, do you know what the/an actual scenario that called for this requirement was?

File: src/handlers/records-read.ts
Comment (diehuxx) [reply src/handlers/records-read.ts:43]: I was expecting the latest created to be returned tbh. What's the rationale for returning the latest updated vs latest created?

@csuwildcat What say you? To borrow/adapt one of your examples, I imagine someone has a blog that they add new blog posts to. They want readers to be able to `RecordsRead` `protocolPath: 'blog/post'` to get the latest created blog post. If the blog owner makes a spelling fix to an older post, they wouldn't want that older post to be returned.

File: tests/handlers/records-read.spec.ts
Comment (diehuxx) [reply tests/handlers/records-read.spec.ts:1957]: Filter's type is `[property: string]: EqualFilter | OneOfFilter | RangeFilter`, so there is typing of the value. For example, the following does not build because `{ bar: { baz: 1 } }` is not a `EqualFilter | OneOfFilter | RangeFilter`

```ts
const filter: Filter = {
  foo: { bar: { baz: 1 } },
};
```

File: src/handlers/records-read.ts
Comment (diehuxx) [reply src/handlers/records-read.ts:92]: Word, makes sense. Can you make it private?

File: tests/handlers/records-read.spec.ts
Comment (diehuxx) [inline tests/handlers/records-read.spec.ts:689]:
```suggestion
            // Install protocol on Alice's DWN
```
Since this the `email` protocol fixture, not `social-media`

**Comment by LiranCohen** [comment]
After today's office hours call we hammered out some final details for this PR and agreed on the following:

`RecordsRead` will now take a `filter` with a `RecordsFilter` type within its descriptor. If that filter returns more than one result it will error back to the user, otherwise it will return the single result.

@thehenrytsai @diehuxx this is ready to be re-reviewed.

File: src/handlers/records-read.ts
Comment (diehuxx) [inline src/handlers/records-read.ts:51]: Since we have `isLatestBaseState: true`, I think we can optimize by doing `if existingMessages.length > 1 then error`.

File: src/handlers/records-read.ts
Comment (diehuxx) [inline src/handlers/records-read.ts:58]: Another optimization as a result of filtering by `isLatestBaseState: true`
```suggestion
    if (newestExistingMessage === undefined) {
```

File: src/handlers/records-read.ts
Comment (diehuxx) [inline src/handlers/records-read.ts:53]: I've been wondering myself what status code we should send. Why 400? My gut tells me this isn't a "bad" request, but it's definitely not a 401. 404 also seems not quite right. Maybe 418? I'll put my musings aside and trust your judgement that 400 is good here. Just want to have it written somewhere if there's a reason.

**Review by diehuxx — COMMENTED** [review]
I like this design a lot better :). Just a couple more small things then we can merge.

File: src/handlers/records-read.ts
Comment (LiranCohen) [reply src/handlers/records-read.ts:53]: 🤣 yeah I had the same thoughts, although not 418 haha!

I landed on 400 because because it feels like a request error from the user's end. They provided some sort of incorrect filter that caused it.

I was thinking 404 at one point, but it felt not descriptive enough since there is a record there, just not only 1.

File: src/handlers/records-read.ts
Comment (LiranCohen) [reply src/handlers/records-read.ts]: Great catch! That actually simplified a lot more here. 🙏

**Comment by LiranCohen** [comment]
@diehuxx Yeah, I'm actually really happy where we landed with this design! It took a little to grow on me, but I think we made the right choices here.

File: src/handlers/records-read.ts
Comment (thehenrytsai) [inline src/handlers/records-read.ts:40]:
```suggestion
    // get the latest active messages matching the supplied filter
```

File: src/handlers/records-read.ts
Comment (thehenrytsai) [inline src/handlers/records-read.ts:55]:
```suggestion
        'Multiple records exist for the RecordsRead filter'
```

File: src/core/dwn-error.ts
Comment (thehenrytsai) [inline src/core/dwn-error.ts:55]: Super nit: should be line 54. 🤣

**Review by thehenrytsai — COMMENTED** [review]
Looks good to me! I got nothing for ya except a few nits, then we get a final stamp of approval for me!

**Comment by LiranCohen** [comment]
Alright, I think this is ready unless there's anything else that was missed! 🚀

File: src/handlers/records-read.ts
Comment (thehenrytsai) [inline src/handlers/records-read.ts:55]:
```suggestion
        'Multiple records exist for the RecordsRead filter'
```

**Review by thehenrytsai — APPROVED** [review]
🎆 🍾 🎖️ 🚀 🎸

**Review by thehenrytsai — APPROVED** [review]

**Review by diehuxx — APPROVED** [review]

**Pull request merged**
